### ISSUE TITLE
Push back on HTTP requests to busy dedicated master nodes

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#70435]
Today we recommend against sending client requests to dedicated master nodes, encouraging users to leave these nodes to perform their duties as master without any unnecessary extra load. However we are very lenient on this front, allowing any node to attempt to coordinate the processing of any incoming request, and users do often send requests to dedicated master nodes either due to misunderstanding or because it’s the most convenient thing to do. This puts the health of the cluster at risk.

I think we should push back on HTTP requests handled by dedicated master nodes and propose that we return a 429 response if we consider ourselves to be too busy with master-related activities to handle work that should be handled by the other nodes. For instance, we might return 429s while we are computing/publishing/applying a new cluster state, or maybe base the decision on the number of pending tasks in the master queue.

I am undecided on which requests should be subject to this new behaviour. Indexing, searches, and stats requests are common sources of problems, but I think it’s reasonable to apply it to ~all requests~ (edit) almost all requests, see [below](https://github.com/elastic/elasticsearch/issues/70435#issuecomment-808020968): if the node is part of a cluster then you can interact with it via the other nodes, and if it’s not part of a cluster then it won’t be busy with cluster state updates so will respond normally.

This proposal is only about dedicated master nodes as found in larger clusters. Master-eligible nodes that have other roles would continue to accept all requests.

This proposal is kinda sorta a breaking change. It’s already technically ok to return a 429 to any request, and clients should handle this appropriately, but this change would make them much more frequent in the cases of master misuse that appear to be common today. I therefore think we should treat it as a breaking change, deprecating the use of busy dedicated masters in 7.x and only enforcing the rule by default in 8.0.

One example of the kind of setup that this proposal would break is Metricbeat’s default “node-scoped” configuration which retrieves most of its stats directly from the elected master. Metricbeat does not need to use the elected master, but it does need some kind of leader selection process: today’s behaviour co-opts the Elasticsearch master election for that purpose. Metricbeat users with larger clusters are expected to configure [“cluster-scoped” monitoring](https://www.elastic.co/guide/en/beats/metricbeat/7.11/metricbeat-module-elasticsearch.html#_module_specific_configuration_notes_5) and use something external for leader selection, but it’s common for users not to do this. Pushing back on stats requests to the master, if it is a dedicated master, would guide these users towards a better setup for their cluster.

In a separate channel we’re discussing some concerns that cluster-scoped monitoring is too difficult to orchestrate. If we conclude that it doesn’t fit our needs then we will need to work with the Metricbeat folks to come up with a different way to organise the Beats to collect cluster-wide metrics which is independent of the Elasticsearch master. That won’t affect the end result of this proposal, but it does add some prerequisites.

It would be good to hear some opinions from the community about the impact of this proposal too.

**Comment by inqueue** [comment]
How about a flag to block heavy request loads entirely to dedicated masters like search and indexing? In this case, the master would respond with an informative 5xx error indicating the node is not accepting such requests.

How in informative would a 429 be? I could see it being difficult to diagnose the cause of 429s if it isn't realized requests are going to dedicated masters.

**Comment by DaveCTurner** [comment]
Thanks @inqueue, yes that's another alternative, but it doesn't protect against all the other things that get sent to dedicated masters in error (e.g. for monitoring). Blocking these completely is also possible, but makes diagnosis of a broken master much harder.

**Comment by DaveCTurner** [comment]
A colleague observed that (in addition) it would be good if we could help clients to avoid dedicated masters automatically where possible: e.g. if a client is configured with a list of all nodes in the cluster then it could reasonably expect each node to expose enough of an API to reliably filter out the dedicated masters from that list.

**Comment by DaveCTurner** [comment]
We (the @elastic/es-distributed team) discussed this yesterday in our team sync. We did not reach any definite conclusions but we did explore some alternative ideas. There seemed to be stronger support for specifically rejecting search and indexing than for rejecting almost everything, and we considered other ways to protect small dedicated masters from overload.

Relates https://github.com/elastic/elasticsearch/issues/51992 since stats requests are a significant load and today we do not have a way to push back on them to make room for master-related activities.
